File: database-backend/src/redis-cache.ts
import { createClient } from 'redis';
import prisma from './prisma';

// Redis client singleton
let redisClient: ReturnType<typeof createClient> | null = null;

/**
 * Initialize Redis connection
 */
export async function initRedis() {
  if (!redisClient) {
    redisClient = createClient({
      url: process.env.REDIS_URL || 'redis://localhost:6379',
      socket: {
        reconnectStrategy: (retries) => Math.min(retries * 50, 500)
      }
    });

    redisClient.on('error', (err) => console.error('Redis Client Error:', err));
    redisClient.on('connect', () => console.log('✅ Redis connected'));
    
    await redisClient.connect();
  }
  return redisClient;
}

/**
 * Get Redis client instance
 */
export function getRedisClient() {
  if (!redisClient) {
    throw new Error('Redis client not initialized. Call initRedis() first.');
  }
  return redisClient;
}

/**
 * Close Redis connection
 */
export async function closeRedis() {
  if (redisClient) {
    await redisClient.quit();
    redisClient = null;
  }
}

// -------------------------------------------
// CACHE KEY PATTERNS
// -------------------------------------------

const CACHE_KEYS = {
  // Grading averages
  COURSE_AVG_SCORE: (courseId: number) => `course:${courseId}:avg_score`,
  COURSE_AVG_POINTS: (courseId: number) => `course:${courseId}:avg_points`,
  TOPIC_AVG_SCORE: (topicId: number) => `topic:${topicId}:avg_score`,
  
  // User-specific comparisons
  USER_COURSE_SCORE: (userId: number, courseId: number) => `user:${userId}:course:${courseId}:score`,
  USER_TOPIC_SCORE: (userId: number, topicId: number) => `user:${userId}:topic:${topicId}:score`,
  
  // Interaction metrics
  COURSE_AVG_INTERACTION_TIME: (courseId: number) => `course:${courseId}:avg_interaction_time`,
  COURSE_AVG_INTERACTION_COUNT: (courseId: number) => `course:${courseId}:avg_interaction_count`,
  
  // Dashboard aggregates (sorted sets for leaderboards)
  COURSE_LEADERBOARD: (courseId: number) => `leaderboard:course:${courseId}`,
  TOPIC_PERFORMANCE: (topicId: number) => `performance:topic:${topicId}`,
  
  // Time-based stats
  DAILY_STATS: (date: string, courseId: number) => `stats:daily:${date}:course:${courseId}`,
};

// Cache TTL (Time To Live) in seconds
const CACHE_TTL = {
  SHORT: 300,      // 5 minutes - for frequently changing data
  MEDIUM: 1800,    // 30 minutes - for moderately stable data
  LONG: 3600,      // 1 hour - for stable aggregates
  DAILY: 86400,    // 24 hours - for daily stats
};

// -------------------------------------------
// GRADING AGGREGATE FUNCTIONS
// -------------------------------------------

/**
 * Calculate and cache average points achieved across all students in a course
 */
export async function getCourseAveragePoints(courseId: number, forceRefresh = false): Promise<number> {
  const client = getRedisClient();
  const cacheKey = CACHE_KEYS.COURSE_AVG_POINTS(courseId);
  
  if (!forceRefresh) {
    const cached = await client.get(cacheKey);
    if (cached) {
      console.log(`📦 Cache hit: ${cacheKey}`);
      return parseFloat(cached);
    }
  }
  
  console.log(`🔍 Cache miss: ${cacheKey} - Querying database...`);
  
  // Calculate average from database
  const result = await prisma.gradingData.aggregate({
    _avg: {
      PointsAchieved: true
    },
    where: {
      // You can add course filtering here if you add course_id to GradingData
    }
  });
  
  const avgPoints = result._avg.PointsAchieved?.toNumber() || 0;
  
  // Cache the result
  await client.setEx(cacheKey, CACHE_TTL.MEDIUM, avgPoints.toString());
  
  return avgPoints;
}

/**
 * Calculate and cache average answer quality score per course
 */
export async function getCourseAverageQualityScore(courseId: number, forceRefresh = false): Promise<number> {
  const client = getRedisClient();
  const cacheKey = CACHE_KEYS.COURSE_AVG_SCORE(courseId);
  
  if (!forceRefresh) {
    const cached = await client.get(cacheKey);
    if (cached) {
      console.log(`📦 Cache hit: ${cacheKey}`);
      return parseFloat(cached);
    }
  }
  
  console.log(`🔍 Cache miss: ${cacheKey} - Querying database...`);
  
  const result = await prisma.gradingData.aggregate({
    _avg: {
      AnswerQualityScore: true
    }
  });
  
  const avgScore = result._avg.AnswerQualityScore?.toNumber() || 0;
  await client.setEx(cacheKey, CACHE_TTL.MEDIUM, avgScore.toString());
  
  return avgScore;
}

/**
 * Get individual student's score and compare with course average
 */
export async function getStudentVsAverageComparison(userId: number, courseId: number) {
  const client = getRedisClient();
  
  // Get student's score (with caching)
  const studentCacheKey = CACHE_KEYS.USER_COURSE_SCORE(userId, courseId);
  let studentScore: number;
  
  const cachedStudentScore = await client.get(studentCacheKey);
  if (cachedStudentScore) {
    studentScore = parseFloat(cachedStudentScore);
  } else {
    // Calculate student's average score
    const studentResult = await prisma.gradingData.aggregate({
      _avg: {
        AnswerQualityScore: true
      },
      where: {
        // Add user_id filter when you add it to GradingData model
      }
    });
    
    studentScore = studentResult._avg.AnswerQualityScore?.toNumber() || 0;
    await client.setEx(studentCacheKey, CACHE_TTL.SHORT, studentScore.toString());
  }
  
  // Get course average
  const courseAverage = await getCourseAverageQualityScore(courseId);
  
  return {
    studentScore,
    courseAverage,
    percentile: ((studentScore / courseAverage) * 100).toFixed(2),
    difference: (studentScore - courseAverage).toFixed(2),
    aboveAverage: studentScore > courseAverage
  };
}

// -------------------------------------------
// INTERACTION AGGREGATE FUNCTIONS
// -------------------------------------------

/**
 * Calculate and cache average interaction time for a course
 */
export async function getCourseAverageInteractionTime(courseId: number, forceRefresh = false): Promise<number> {
  const client = getRedisClient();
  const cacheKey = CACHE_KEYS.COURSE_AVG_INTERACTION_TIME(courseId);
  
  if (!forceRefresh) {
    const cached = await client.get(cacheKey);
    if (cached) {
      console.log(`📦 Cache hit: ${cacheKey}`);
      return parseFloat(cached);
    }
  }
  
  console.log(`🔍 Cache miss: ${cacheKey} - Querying database...`);
  
  const result = await prisma.interactionData.aggregate({
    _avg: {
      DurationSeconds: true
    }
  });
  
  const avgTime = result._avg.DurationSeconds || 0;
  await client.setEx(cacheKey, CACHE_TTL.MEDIUM, avgTime.toString());
  
  return avgTime;
}

/**
 * Calculate and cache average interaction count per session
 */
export async function getCourseAverageInteractionCount(courseId: number, forceRefresh = false): Promise<number> {
  const client = getRedisClient();
  const cacheKey = CACHE_KEYS.COURSE_AVG_INTERACTION_COUNT(courseId);
  
  if (!forceRefresh) {
    const cached = await client.get(cacheKey);
    if (cached) return parseInt(cached);
  }
  
  const result = await prisma.interactionData.aggregate({
    _avg: {
      InteractionCount: true
    }
  });
  
  const avgCount = result._avg.InteractionCount || 0;
  await client.setEx(cacheKey, CACHE_TTL.MEDIUM, avgCount.toString());
  
  return avgCount;
}

// -------------------------------------------
// LEADERBOARD FUNCTIONS (Sorted Sets)
// -------------------------------------------

/**
 * Update course leaderboard with student scores
 * Uses Redis Sorted Sets for efficient ranking
 */
export async function updateCourseLeaderboard(courseId: number) {
  const client = getRedisClient();
  const leaderboardKey = CACHE_KEYS.COURSE_LEADERBOARD(courseId);
  
  console.log(`🏆 Updating leaderboard for course ${courseId}...`);
  
  // Get all students' scores (you'll need to join with user data)
  // This is a simplified example - adjust based on your schema
  const scores = await prisma.gradingData.groupBy({
    by: ['id'], // Replace with user_id when available
    _avg: {
      AnswerQualityScore: true
    }
  });
  
  // Update sorted set
  for (const score of scores) {
    const avgScore = score._avg.AnswerQualityScore?.toNumber() || 0;
    await client.zAdd(leaderboardKey, {
      score: avgScore,
      value: `student:${score.id}` // Replace with actual user_id
    });
  }
  
  // Set expiration
  await client.expire(leaderboardKey, CACHE_TTL.LONG);
  
  console.log(`✅ Leaderboard updated with ${scores.length} entries`);
}

/**
 * Get top N students from leaderboard
 */
export async function getTopStudents(courseId: number, limit = 10) {
  const client = getRedisClient();
  const leaderboardKey = CACHE_KEYS.COURSE_LEADERBOARD(courseId);
  
  // Get top scorers (descending order)
  const topStudents = await client.zRangeWithScores(leaderboardKey, 0, limit - 1, {
    REV: true // Reverse order for highest scores first
  });
  
  return topStudents.map((entry, index) => ({
    rank: index + 1,
    studentId: entry.value,
    score: entry.score
  }));
}

/**
 * Get student's rank in course
 */
export async function getStudentRank(courseId: number, studentId: number) {
  const client = getRedisClient();
  const leaderboardKey = CACHE_KEYS.COURSE_LEADERBOARD(courseId);
  
  const rank = await client.zRevRank(leaderboardKey, `student:${studentId}`);
  const score = await client.zScore(leaderboardKey, `student:${studentId}`);
  
  return {
    rank: rank !== null ? rank + 1 : null, // Convert to 1-based ranking
    score: score || 0
  };
}

// -------------------------------------------
// CACHE INVALIDATION
// -------------------------------------------

/**
 * Invalidate all caches related to a course
 */
export async function invalidateCourseCaches(courseId: number) {
  const client = getRedisClient();
  
  const keysToDelete = [
    CACHE_KEYS.COURSE_AVG_SCORE(courseId),
    CACHE_KEYS.COURSE_AVG_POINTS(courseId),
    CACHE_KEYS.COURSE_AVG_INTERACTION_TIME(courseId),
    CACHE_KEYS.COURSE_AVG_INTERACTION_COUNT(courseId),
    CACHE_KEYS.COURSE_LEADERBOARD(courseId),
  ];
  
  await client.del(keysToDelete);
  console.log(`🗑️  Invalidated ${keysToDelete.length} cache keys for course ${courseId}`);
}

/**
 * Invalidate user-specific caches
 */
export async function invalidateUserCaches(userId: number) {
  const client = getRedisClient();
  
  // Find all keys matching pattern user:{userId}:*
  const keys = await client.keys(`user:${userId}:*`);
  
  if (keys.length > 0) {
    await client.del(keys);
    console.log(`🗑️  Invalidated ${keys.length} cache keys for user ${userId}`);
  }
}

/**
 * Clear all caches (use with caution!)
 */
export async function clearAllCaches() {
  const client = getRedisClient();
  await client.flushDb();
  console.log('🗑️  All caches cleared');
}

// -------------------------------------------
// BULK OPERATIONS
// -------------------------------------------

/**
 * Precompute and cache all course statistics
 * Run this periodically (e.g., every hour via cron job)
 */
export async function precomputeCourseStats(courseId: number) {
  console.log(`⚙️  Precomputing stats for course ${courseId}...`);
  
  await Promise.all([
    getCourseAveragePoints(courseId, true),
    getCourseAverageQualityScore(courseId, true),
    getCourseAverageInteractionTime(courseId, true),
    getCourseAverageInteractionCount(courseId, true),
    updateCourseLeaderboard(courseId),
  ]);
  
  console.log(`✅ Stats precomputed for course ${courseId}`);
}

/**
 * Export for dashboard API
 */
export async function getDashboardStats(courseId: number, userId?: number) {
  const courseStats = {
    averagePoints: await getCourseAveragePoints(courseId),
    averageQualityScore: await getCourseAverageQualityScore(courseId),
    averageInteractionTime: await getCourseAverageInteractionTime(courseId),
    averageInteractionCount: await getCourseAverageInteractionCount(courseId),
    topStudents: await getTopStudents(courseId, 10),
  };
  
  let userComparison;
  if (userId) {
    userComparison = await getStudentVsAverageComparison(userId, courseId);
  }
  
  return {
    course: courseStats,
    user: userComparison,
    timestamp: new Date().toISOString(),
  };
}

File: python-backend/redis_cache.py
"""
Redis caching layer for dashboard statistics and metrics.
Implements caching strategies for frequently accessed aggregated data.
"""

import os
import json
from typing import Optional, Dict, Any, List
from redis import Redis
from dotenv import load_dotenv
from sqlalchemy import func
from database import get_db_session
from models import (
    GradingData, InteractionData, User, Course, 
    Chatbot, Conversation, Message, Question, Answer
)

# Load environment variables
load_dotenv()

# Redis client singleton
_redis_client: Optional[Redis] = None


def get_redis_client() -> Redis:
    """Get or create Redis client instance"""
    global _redis_client
    
    if _redis_client is None:
        redis_url = os.environ.get('REDIS_URL', 'redis://localhost:6379')
        _redis_client = Redis.from_url(
            redis_url,
            decode_responses=True,
            socket_connect_timeout=5,
            socket_keepalive=True
        )
    
    return _redis_client


def init_redis():
    """Initialize and test Redis connection"""
    try:
        client = get_redis_client()
        client.ping()
        print("✅ Redis connected successfully!")
        return client
    except Exception as e:
        print(f"❌ Redis connection failed: {e}")
        raise


def close_redis():
    """Close Redis connection"""
    global _redis_client
    if _redis_client:
        _redis_client.close()
        _redis_client = None


# Cache key patterns
class CacheKeys:
    COURSE_AVG_SCORE = "course:{course_id}:avg_score"
    COURSE_AVG_POINTS = "course:{course_id}:avg_points"
    USER_COURSE_SCORE = "user:{user_id}:course:{course_id}:score"
    COURSE_AVG_INTERACTION_TIME = "course:{course_id}:avg_interaction_time"
    COURSE_AVG_INTERACTION_COUNT = "course:{course_id}:avg_interaction_count"
    COURSE_LEADERBOARD = "leaderboard:course:{course_id}"
    COURSE_STATS = "stats:course:{course_id}"


# Cache TTL (Time To Live) in seconds
class CacheTTL:
    SHORT = 300      # 5 minutes
    MEDIUM = 1800    # 30 minutes
    LONG = 3600      # 1 hour
    DAILY = 86400    # 24 hours


def get_cached(key: str) -> Optional[str]:
    """Get value from cache"""
    try:
        client = get_redis_client()
        return client.get(key)
    except Exception as e:
        print(f"Cache read error: {e}")
        return None


def set_cached(key: str, value: Any, ttl: int = CacheTTL.MEDIUM):
    """Set value in cache with TTL"""
    try:
        client = get_redis_client()
        if isinstance(value, (dict, list)):
            value = json.dumps(value)
        client.setex(key, ttl, str(value))
    except Exception as e:
        print(f"Cache write error: {e}")


def invalidate_cache(pattern: str):
    """Invalidate cache keys matching pattern"""
    try:
        client = get_redis_client()
        keys = client.keys(pattern)
        if keys:
            client.delete(*keys)
            print(f"🗑️ Invalidated {len(keys)} cache keys")
    except Exception as e:
        print(f"Cache invalidation error: {e}")


# Dashboard statistics functions

def get_course_average_points(course_id: int, force_refresh: bool = False) -> float:
    """Calculate and cache average points achieved in a course"""
    cache_key = CacheKeys.COURSE_AVG_POINTS.format(course_id=course_id)
    
    if not force_refresh:
        cached = get_cached(cache_key)
        if cached:
            print(f"📦 Cache hit: {cache_key}")
            return float(cached)
    
    print(f"🔍 Cache miss: {cache_key} - Querying database...")
    db = get_db_session()
    try:
        result = db.query(func.avg(GradingData.PointsAchieved)).scalar()
        avg_points = float(result) if result else 0.0
        set_cached(cache_key, avg_points, CacheTTL.MEDIUM)
        return avg_points
    finally:
        db.close()


def get_course_average_quality_score(course_id: int, force_refresh: bool = False) -> float:
    """Calculate and cache average answer quality score for a course"""
    cache_key = CacheKeys.COURSE_AVG_SCORE.format(course_id=course_id)
    
    if not force_refresh:
        cached = get_cached(cache_key)
        if cached:
            print(f"📦 Cache hit: {cache_key}")
            return float(cached)
    
    print(f"🔍 Cache miss: {cache_key} - Querying database...")
    db = get_db_session()
    try:
        result = db.query(func.avg(GradingData.AnswerQualityScore)).scalar()
        avg_score = float(result) if result else 0.0
        set_cached(cache_key, avg_score, CacheTTL.MEDIUM)
        return avg_score
    finally:
        db.close()


def get_course_average_interaction_time(course_id: int, force_refresh: bool = False) -> int:
    """Calculate and cache average interaction time for a course"""
    cache_key = CacheKeys.COURSE_AVG_INTERACTION_TIME.format(course_id=course_id)
    
    if not force_refresh:
        cached = get_cached(cache_key)
        if cached:
            print(f"📦 Cache hit: {cache_key}")
            return int(float(cached))
    
    print(f"🔍 Cache miss: {cache_key} - Querying database...")
    db = get_db_session()
    try:
        result = db.query(func.avg(InteractionData.DurationSeconds)).scalar()
        avg_time = int(result) if result else 0
        set_cached(cache_key, avg_time, CacheTTL.MEDIUM)
        return avg_time
    finally:
        db.close()


def get_student_vs_average_comparison(user_id: int, course_id: int) -> Dict[str, Any]:
    """Compare individual student's score with course average"""
    cache_key = CacheKeys.USER_COURSE_SCORE.format(user_id=user_id, course_id=course_id)
    
    # Get student score (implement based on your logic)
    cached_student_score = get_cached(cache_key)
    if cached_student_score:
        student_score = float(cached_student_score)
    else:
        # Calculate student's average score from grading data
        db = get_db_session()
        try:
            # This is a placeholder - adjust based on how you link users to grading data
            result = db.query(func.avg(GradingData.AnswerQualityScore)).scalar()
            student_score = float(result) if result else 0.0
            set_cached(cache_key, student_score, CacheTTL.SHORT)
        finally:
            db.close()
    
    # Get course average
    course_average = get_course_average_quality_score(course_id)
    
    return {
        'studentScore': student_score,
        'courseAverage': course_average,
        'percentile': round((student_score / course_average * 100), 2) if course_average > 0 else 0,
        'difference': round(student_score - course_average, 2),
        'aboveAverage': student_score > course_average
    }


def update_course_leaderboard(course_id: int):
    """Update course leaderboard using Redis sorted sets"""
    leaderboard_key = CacheKeys.COURSE_LEADERBOARD.format(course_id=course_id)
    
    print(f"🏆 Updating leaderboard for course {course_id}...")
    db = get_db_session()
    client = get_redis_client()
    
    try:
        # Get all students' scores (adjust query based on your schema)
        # This is a placeholder - you'll need to join properly with user data
        results = db.query(
            GradingData.id,
            func.avg(GradingData.AnswerQualityScore).label('avg_score')
        ).group_by(GradingData.id).all()
        
        # Update sorted set
        for result in results:
            client.zadd(leaderboard_key, {f"student:{result.id}": float(result.avg_score)})
        
        # Set expiration
        client.expire(leaderboard_key, CacheTTL.LONG)
        print(f"✅ Leaderboard updated with {len(results)} entries")
    finally:
        db.close()


def get_top_students(course_id: int, limit: int = 10) -> List[Dict[str, Any]]:
    """Get top N students from leaderboard"""
    leaderboard_key = CacheKeys.COURSE_LEADERBOARD.format(course_id=course_id)
    client = get_redis_client()
    
    top_students = client.zrevrange(leaderboard_key, 0, limit - 1, withscores=True)
    
    return [
        {
            'rank': index + 1,
            'studentId': student_id.decode() if isinstance(student_id, bytes) else student_id,
            'score': score
        }
        for index, (student_id, score) in enumerate(top_students)
    ]


def get_student_rank(course_id: int, user_id: int) -> Dict[str, Any]:
    """Get student's rank in course leaderboard"""
    leaderboard_key = CacheKeys.COURSE_LEADERBOARD.format(course_id=course_id)
    client = get_redis_client()
    
    student_key = f"student:{user_id}"
    rank = client.zrevrank(leaderboard_key, student_key)
    score = client.zscore(leaderboard_key, student_key)
    
    return {
        'rank': rank + 1 if rank is not None else None,
        'score': score if score else 0.0
    }


def precompute_course_stats(course_id: int):
    """Precompute and cache all statistics for a course"""
    print(f"⚙️ Precomputing stats for course {course_id}...")
    
    get_course_average_points(course_id, force_refresh=True)
    get_course_average_quality_score(course_id, force_refresh=True)
    get_course_average_interaction_time(course_id, force_refresh=True)
    update_course_leaderboard(course_id)
    
    print(f"✅ Stats precomputed for course {course_id}")


def get_dashboard_stats(course_id: int, user_id: Optional[int] = None) -> Dict[str, Any]:
    """Get complete dashboard statistics for a course"""
    course_stats = {
        'averagePoints': get_course_average_points(course_id),
        'averageQualityScore': get_course_average_quality_score(course_id),
        'averageInteractionTime': get_course_average_interaction_time(course_id),
        'topStudents': get_top_students(course_id, 10)
    }
    
    user_comparison = None
    if user_id:
        user_comparison = get_student_vs_average_comparison(user_id, course_id)
    
    return {
        'course': course_stats,
        'user': user_comparison,
        'timestamp': str(datetime.now())
    }


def invalidate_course_caches(course_id: int):
    """Invalidate all caches related to a course"""
    patterns = [
        f"course:{course_id}:*",
        f"leaderboard:course:{course_id}",
        f"stats:course:{course_id}"
    ]
    
    for pattern in patterns:
        invalidate_cache(pattern)


if __name__ == "__main__":
    from datetime import datetime
    
    # Test Redis connection
    print("Testing Redis connection...")
    try:
        init_redis()
        
        # Test basic cache operations
        print("\nTesting cache operations...")
        set_cached("test:key", "test_value", 60)
        value = get_cached("test:key")
        print(f"Cached value: {value}")
        
        print("\n✅ Redis cache module is ready!")
    except Exception as e:
        print(f"\n❌ Redis test failed: {e}")
    finally:
        close_redis()
